### DISCUSSION TITLE
Native iOS Version  - Discussion Post

### DISCUSSION THREAD

**Discussion by Danielku15** [CoderLine/alphaTab#1332] [General]
You can use this discussion post to ask questions and discuss things around the announcement we made [here](https://github.com/CoderLine/alphaTab/issues/1333).

**Comment by zhouhao27** [comment]
I'm wondering if still somebody interested at this topic. I want to use alphaTab in Apple TV but unfortunately there is no web view supported in Apple TV then we have no way to use alphaTab In Apple TV. The native support is the only choice.

**Reply by Danielku15** [reply]
That's quite unfortunate. Depending on how open you are with your tech-stack, one option could be to use .net technologies like MAUI, Avalonia or Uno. 

We do not have (yet) out-of-the-box controls for these technologies but it's something I could look into. 

It's quite unfortunate with the iOS ecosystem (as described in the discussion post) and I have no plans looking into an iOS version unless it is a platform where I can develop outside a Mac (and it's free to develop).

**Reply by zhouhao27** [reply]
> That's quite unfortunate. Depending on how open you are with your tech-stack, one option could be to use .net technologies like MAUI, Avalonia or Uno.
> 
> We do not have (yet) out-of-the-box controls for these technologies but it's something I could look into.
> 
> It's quite unfortunate with the iOS ecosystem (as described in the discussion post) and I have no plans looking into an iOS version unless it is a platform where I can develop outside a Mac (and it's free to develop).

Thanks for your reply. I went through the source code of alphaTab.kotline. I think it converts the TypeScript/JavaScript into Kotlin code when you build the library and create renderer through Skia with kotline. Is it possible to do the same thing for iOS? Make you can provide some hints for me. I'll do further research. Thanks.

**Reply by Danielku15** [reply]
> Is it possible to do the same thing for iOS? Make you can provide some hints for me. 

You can read the details in the announcement post linked above: https://github.com/CoderLine/alphaTab/issues/1333

Technically another challenge is that iOS with Swift/C++/ObjectiveC does not have a garbage collection (unlike all other supported platforms). We always need to target some environment serving this need (Kotlin/Native or KotlinMPP, .net Native AOT, etc.).

Eventually we need to properly maintain the iOS platform (extending features whenever needed, providing builds/packages etc.) A one time contribution to the project might bring us up to the basics but its not sustainable. 

The iOS ecosystem is basically a dictatorship run by Apple aiming hard to avoid that people can develop outside their defined borders. Unless you buy/rent a Mac, and pay 100$/yr for a certificate its basically not possible to contribute to the community. 

If you want to dig down this rabbit hole you could look into Kotlin MPP / Kotlin Native or .net native AOT

* https://kotlinlang.org/docs/multiplatform.html
* https://github.com/CoderLine/alphaTab/tree/feature/native-win-x64
* https://kotlinlang.org/docs/native-overview.html
* https://github.com/dotnet/runtime/discussions/100120
* https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=windows%2Cnet8#platformarchitecture-restrictions
* https://github.com/dotnet/runtime/tree/main/src/mono/sample/iOS-NativeAOT

**Comment by zhouhao27** [comment]
I noticed you built a library called `alphaskia` which supports iOS build.And `skia` itself seems support iOS too. I'm wondering if it's possible to build a native iOS version from `alphaskia` or `skia` instead of from kotlin.

**Reply by Danielku15** [reply]
alphaSkia is a compilation of Skia hidden behind some wrapper. It is for different platforms including .net iOS versions.

https://github.com/CoderLine/alphaSkia?tab=readme-ov-file#supported-platforms

It is not a general "iOS" library and not comparable with alphaTab in terms of compilation and it is basically untested on iOS. The core part of this library is C/C++.

If you actually mean: "is it possible to add a Kotlin/Native build to alphaSkia to have it available on iOS". I guess this would be possible.
